src/lib/rnp.cpp

```diff
@@ -6175,12 +6168,72 @@ try {
     if (idx >= key->subkey_count()) {
         return RNP_ERROR_BAD_PARAMETERS;
     }
-    const pgp_fingerprint_t &fp = key->get_subkey_fp(idx);
-    char                     fphex[PGP_FINGERPRINT_SIZE * 2 + 1] = {0};
-    if (!rnp::hex_encode(fp.fingerprint, fp.length, fphex, sizeof(fphex))) {
-        return RNP_ERROR_BAD_STATE;
+    pgp_key_search_t locator = {};
+    locator.type = PGP_KEY_SEARCH_FINGERPRINT;
+    locator.by.fingerprint = key->get_subkey_fp(idx);
+    return rnp_locate_key_int(handle->ffi, locator, subkey);
+}
+FFI_GUARD
+
+rnp_result_t
+rnp_key_get_default_key(rnp_key_handle_t  primary_key,
+                        const char *      usage,
+                        uint32_t          flags,
+                        rnp_key_handle_t *default_key)
+try {
+    if (!primary_key || !usage || !default_key) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    uint8_t keyflag = 0;
+    bool    no_primary = false;
+    if (!str_to_key_flag(usage, &keyflag)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    if (flags & RNP_KEY_SUBKEYS_ONLY) {
+        no_primary = true;
+        flags &= ~RNP_KEY_SUBKEYS_ONLY;
     }
-    return rnp_locate_key(handle->ffi, "fingerprint", fphex, subkey);
+    if (flags) {
+        FFI_LOG(primary_key->ffi, "Invalid flags: %" PRIu32, flags);
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    pgp_key_t *key = get_key_prefer_public(primary_key);
+    if (!key) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    pgp_key_t *defkey = find_suitable_key(
+      PGP_OP_UNKNOWN, key, &primary_key->ffi->key_provider, keyflag, no_primary);
+    if (!defkey) {
+        *default_key = NULL;
+        return RNP_ERROR_NO_SUITABLE_KEY;
+    }
+    pgp_key_search_t search = {(pgp_key_search_type_t) 0};
+    search.type = PGP_KEY_SEARCH_FINGERPRINT;
+    search.by.fingerprint = defkey->fp();
+
+    // search pubring
+    pgp_key_t *pub = rnp_key_store_search(primary_key->ffi->pubring, &search, NULL);
+    // search secring
+    pgp_key_t *sec = rnp_key_store_search(primary_key->ffi->secring, &search, NULL);
+
+    if (!sec && keyflag != PGP_KF_ENCRYPT) {
+        return RNP_ERROR_NO_SUITABLE_KEY;
+    }
+
+    if (pub || sec) {
```

Comment:
We can move pubring search here

Reply:
@rrrooommmaaa This was extracted to this issue: https://github.com/rnpgp/rnp/issues/1570
Actually, as release only cherry-picks commits from the master branch, we could not change code here, as first we'll need to create issues/PRs for the master branch.

Reply:
Sure, as you see fit